bundles/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/handler/ICloudAccountBridgeHandler.java

```diff
@@ -88,85 +101,281 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+        if (authState == AuthState.WAIT_FOR_CODE) {
+            throw new IllegalStateException("2-FA authentication not completed.");
+        }
+
         try {
-            logger.debug("iCloud bridge starting handler ...");
             ICloudAccountThingConfiguration config = getConfigAs(ICloudAccountThingConfiguration.class);
-            final String localAppleId = config.appleId;
-            final String localPassword = config.password;
-            if (localAppleId != null && localPassword != null) {
-                connection = new ICloudConnection(localAppleId, localPassword);
-            } else {
+            boolean success = false;
+
+            // No code requested yet or session is trusted (hopefully).
+            success = this.iCloudService.authenticate(false);
+            if (!success) {
+                authState = AuthState.USER_PW_INVALID;
+                logger.warn("iCloud authentication failed. Invalid credentials.");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Invalid credentials.");
+                this.iCloudService = null;
+                return false;
+            }
+            if (iCloudService.requires2fa()) {
+                // New code was requested. Wait for the user to update config.
+                logger.warn(
+                        "iCloud authentication requires 2-FA code. Please provide code configuration for thing '{}'.",
+                        getThing().getUID().getAsString());
+                authState = AuthState.WAIT_FOR_CODE;
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                        "Apple ID/Password is not set!");
-                return;
+                        "Please provide 2-FA code in thing configuration.");
+                return false;
+            }
+
+            if (!this.iCloudService.isTrustedSession()) {
+                logger.debug("Trying to establish session trust.");
+                success = this.iCloudService.trustSession();
+                if (!success) {
+                    logger.warn("iCloud trust session failed.");
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Session trust failed.");
+                    return false;
+                }
             }
-            refreshJob = scheduler.scheduleWithFixedDelay(this::refreshData, 0, config.refreshTimeInMinutes, MINUTES);
 
-            logger.debug("iCloud bridge handler started.");
-        } catch (URISyntaxException e) {
-            logger.debug("Something went wrong while constructing the connection object", e);
+            authState = AuthState.AUTHENTICATED;
+            updateStatus(ThingStatus.ONLINE);
+            logger.debug("iCloud bridge handler authenticated.");
+            return true;
+
+        } catch (IOException | InterruptedException e) {
+            logger.warn("iCloud authentication caused exception.", e);
```

Comment:
This could be reduced to **debug** or **trace** level (to have full stack trace), since the framework will already log the exception message from the status change below.